pkg/controllers/openshift_apiserver/controller.go

```diff
@@ -0,0 +1,243 @@
+	panic("informer not supported")
+}
+
+func (c *apiServerOperatorClient) GetOperatorState() (spec *operatorv1.OperatorSpec, status *operatorv1.OperatorStatus, resourceVersion string, err error) {
+	var cm *corev1.ConfigMap
+	cm, err = c.Client.CoreV1().ConfigMaps(c.Namespace).Get(apiserverConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	configYAML := []byte(cm.Data["config.yaml"])
+	var configJSON []byte
+	configJSON, err = yaml.YAMLToJSON(configYAML)
+	if err != nil {
+		return
+	}
+	configJSON, err = filterManagedConfigKeys(configJSON)
+	if err != nil {
+		return
+	}
+	spec = &operatorv1.OperatorSpec{}
+	status = &operatorv1.OperatorStatus{}
+	spec.ObservedConfig.Raw = configJSON
+	resourceVersion = cm.ResourceVersion
+	return
+}
+
+// UpdateOperatorSpec updates the spec of the operator, assuming the given resource version.
+func (c *apiServerOperatorClient) UpdateOperatorSpec(oldResourceVersion string, in *operatorv1.OperatorSpec) (out *operatorv1.OperatorSpec, newResourceVersion string, err error) {
+	var cm *corev1.ConfigMap
+	cm, err = c.Client.CoreV1().ConfigMaps(c.Namespace).Get(apiserverConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	if cm.ResourceVersion != oldResourceVersion {
+		err = fmt.Errorf("resource version does not match")
+		return
+	}
+	var updateJSON []byte
+	updateJSON, err = in.ObservedConfig.MarshalJSON()
+	if err != nil {
+		return
+	}
+	var configBytes []byte
+	configBytes, err = mergeConfig([]byte(cm.Data["config.yaml"]), updateJSON)
+	if err != nil {
+		return
+	}
+	cm.Data["config.yaml"] = string(configBytes)
+	c.Logger.Info("Updating OpenShift APIServer configmap")
+	_, err = c.Client.CoreV1().ConfigMaps(c.Namespace).Update(cm)
+	if err != nil {
+		return
+	}
+	dataHash := calculateHash(configBytes)
+	var deployment *appsv1.Deployment
+	deployment, err = c.Client.AppsV1().Deployments(c.Namespace).Get(openshiftDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	if deployment.Spec.Template.ObjectMeta.Annotations == nil {
+		deployment.Spec.Template.ObjectMeta.Annotations = map[string]string{}
+	}
+	c.Logger.Info("Updating OpenShift APIServer deployment")
+	deployment.Spec.Template.ObjectMeta.Annotations["config-checksum"] = dataHash
+	_, err = c.Client.AppsV1().Deployments(c.Namespace).Update(deployment)
+	return
+}
+
+func mergeConfig(existingYAML, updateJSON []byte) (updatedYAML []byte, err error) {
+	var existingJSON []byte
+	existingJSON, err = yaml.YAMLToJSON(existingYAML)
+	if err != nil {
+		return
+	}
+	existingConfig := map[string]interface{}{}
+	if err = json.NewDecoder(bytes.NewBuffer(existingJSON)).Decode(&existingConfig); err != nil {
+		return
+	}
+	updateConfig := map[string]interface{}{}
+	if err = json.NewDecoder(bytes.NewBuffer(updateJSON)).Decode(&updateConfig); err != nil {
+		return
+	}
+	for key := range updateConfig {
+		switch key {
+		case "imagePolicyConfig", "projectConfig", "routingConfig":
```

Comment:
fundamentally within IKS:
routingConfig if it changes will cause us a load of problems and our microservices will have issues. Can we remove that for now? L

imagePolicyConfig looks to be one that would allow the user to shoot themselves in the foot as well if they changed to a different registry url that doesn't match up with the registry service. Can we remove that as a supported option.

Let me know if I'm missing something though because I might be.

Reply:
Yes routing config lets you change the routing subdomain which I see can be problematic. Removing for now.
The imagePolicyConfig has 3 different observers:

- external registry hostname - yup, can be bad if changed ... will remove
- internal registry hostname - only keeps in sync with the status reported by the image registry operator ... not configurable by the user
- allowed registries for import - the whitelist of registries that [normal users  can import from.](https://github.com/openshift/api/blob/0de0d539b0c32b2f1d7255c3100a7e92df2a99e2/config/v1/types_image.go#L36)  Let me know if this also should be restricted by ROKS.